File: kotlin/src/main/kotlin/155MinStack.kt
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.*

//https://leetcode.com/problems/min-stack/


class MinStack{
  private val mins = Stack<Int>()
  private val values = Stack<Int>()

  fun push(x: Int) {
    if(mins.empty() || x <= mins.peek())
      mins.push(x)
    values.push(x)
  }

  fun pop(): Int {
    val v = values.pop()
    if(mins.peek() == v) mins.pop()
    return v
  }

  fun top(): Int = values.peek()

  fun getMin() = mins.peek()

  @Test
  fun test1(){
    push(-2)
    push(0)
    push(-3)
    assertEquals(-3, getMin())
    pop()
    assertEquals(0, top())
    assertEquals(-2, getMin())
  }
}

File: python/155_MinStack.py
from unittest import TestCase
# https://leetcode.com/problems/min-stack/


class MinStack:
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.mins = []
        self.values = []

    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        if len(self.mins) == 0 or x <= self.mins[-1]:
            self.mins.append(x)
        self.values.append(x)

    def pop(self):
        """
        :rtype: void
        """
        if self.mins[-1] == self.values.pop():
            self.mins.pop()

    def top(self):
        """
        :rtype: int
        """
        return self.values[-1]

    def getMin(self):
        """
        :rtype: int
        """
        return self.mins[-1]


class Tests(TestCase):
    def test1(self):
        s = MinStack()
        s.push(-2)
        s.push(0)
        s.push(-3)
        self.assertEqual(-3, s.getMin())
        s.pop()
        self.assertEqual(0, s.top())
        self.assertEqual(-2, s.getMin())
